core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.controller
+
+import java.util.concurrent.locks.ReentrantLock
+import java.util.concurrent.{ArrayBlockingQueue, ConcurrentHashMap, CountDownLatch}
+
+import kafka.utils.CoreUtils.inLock
+import org.apache.zookeeper.AsyncCallback.{ACLCallback, Children2Callback, DataCallback, StatCallback, StringCallback, VoidCallback}
+import org.apache.zookeeper.Watcher.Event.EventType
+import org.apache.zookeeper.ZooKeeper.States
+import org.apache.zookeeper.data.{ACL, Stat}
+import org.apache.zookeeper.{CreateMode, WatchedEvent, Watcher, ZooKeeper}
+
+class ZookeeperClient(connectString: String, sessionTimeout: Int) {
+  private val initializationLock = new ReentrantLock()
+  private val isConnectedOrExpiredLock = new ReentrantLock()
+  private val isConnectedOrExpiredCondition = isConnectedOrExpiredLock.newCondition()
+  private val zNodeChangeHandlers = new ConcurrentHashMap[String, ZNodeChangeHandler]()
+  private val zNodeChildChangeHandlers = new ConcurrentHashMap[String, ZNodeChildChangeHandler]()
+  @volatile private var stateChangeHandlerOpt: Option[StateChangeHandler] = None
+  @volatile private var zooKeeper = new ZooKeeper(connectString, sessionTimeout, ZookeeperClientWatcher)
+  private val sessionContext = new ThreadLocal[ZooKeeper] {
+    override def initialValue(): ZooKeeper = zooKeeper
```

Comment:
Most ZK accesses are oblivious to which ZK session the access is on. The only exception is the controller logic and the registration of the broker id. So, we can just register a session expiration handler, in which we first clear the controller queue and add a new session expiration event to the controller queue. Then, once the controller thread picks up the session expiration event, it can create a new ZK session, re-register the broker id and do controller reelection. No other threads can create new ZK session. If they get an error on the current session, they can simply wait for the ZK state to be ready again and resume the operation.